### DISCUSSION TITLE
Is it possible to enable threads? - I get `RuntimeError: can't start new thread` when trying to use `rich` `Live`

### DISCUSSION THREAD

**Discussion by goulashsoup** [pyscript/pyscript#2086] [Q&A] [answered]
I have just this very simple setup:

`web_server.py`:

```python
from http.server import SimpleHTTPRequestHandler, HTTPServer

class MyHTTPRequestHandler(SimpleHTTPRequestHandler):
    def end_headers(self):
        self.send_my_headers()

        # Call the original end_headers() method
        SimpleHTTPRequestHandler.end_headers(self)

    def send_my_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "require-corp")
        self.send_header("Cross-Origin-Resource-Policy", "cross-origin")

def run(server_class=HTTPServer, handler_class=MyHTTPRequestHandler, port=8000):
    server_address = ('localhost', port)
    httpd = server_class(server_address, handler_class)
    print(f"Starting server on port {port}...")
    httpd.serve_forever()

if __name__ == "__main__":
    run()
```

`index.html`:

```HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <title>Crimson Credit</title>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="stylesheet" href="https://pyscript.net/releases/2024.5.2/core.css">
    <script type="module" src="https://pyscript.net/releases/2024.5.2/core.js"></script>
</head>
<body>
    <script type="py" src="./main.py" config="./pyscript.toml"></script>
    <!--  Also does NOT work with `worker` attribute:   <script type="py" src="./main.py" config="./pyscript.toml" worker></script> -->
</body>
</html>
```

`pyscript.toml`:

```toml
packages = ["rich"]
```

`main.py`:

```python
from rich.console import Group
from rich.live import Live
import time

output_renderables = ['hello']
outputs = Group(*output_renderables)
live = Live(outputs)
live.start()
time.sleep(5)
```

This gives following error:

```
PythonError: Traceback (most recent call last):
  File "/lib/python311.zip/_pyodide/_base.py", line 501, in eval_code
    .run(globals, locals)
     ^^^^^^^^^^^^^^^^^^^^
  File "/lib/python311.zip/_pyodide/_base.py", line 339, in run
    coroutine = eval(self.code, globals, locals)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<exec>", line 8, in <module>
  File "/lib/python3.11/site-packages/rich/live.py", line 132, in start
    self._refresh_thread.start()
  File "/lib/python311.zip/threading.py", line 957, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
M — pyodide.asm.js:9:86987
new_error — pyodide.asm.js:9:10034
<?>.wasm-function[300]
<?>.wasm-function[301]
(anonymous function) — pyodide.asm.js:9:64422
At — 6948aca8-0378-4896-818b-b469107e06b8:2:11545
(anonymous function) — 6948aca8-0378-4896-818b-b469107e06b8:2:22597
(anonymous function) — 6948aca8-0378-4896-818b-b469107e06b8:2:30023
(anonymous function) — 6948aca8-0378-4896-818b-b469107e06b8:2:31034
```

I think this is related to https://github.com/pyodide/pyodide/issues/237, as pyodide doesn't support threading yet. If so, I had to use a "custom" pyodide version, unfortunately the latest suggested "fix" doesn't work for me either (see latest comment).

I just wonder:  **How much effort would it to get it to work?**

I also found [Monkeypatching Rich for Beautiful Terminals in Pyscript](https://jeff.glass/post/pyscript-rich/), but this just monkeypatches `Live`, so it does NOT fix the threading problem, just works around it.

**Could we "overwrite" the core `_start_new_thread` function to "simulate" threading?**

I also don't understand the problem quiet well, as pyscript already allows WebWorkers and async, I would have expected threading to just "work"?

At last **Where should I "start" getting it to work?**
**What do you think would be the easiest approach?**

I appreciate any constructive feedback, Thank you!

**Comment by WebReflection** [comment, marked as answer]
if threading is not supported in *Pyodide* there's rally very little we can do in here but I think we had FastAPI working out of a Service Worker and doing a good job there ... it requires you bootstrapping pyodide in there though, as we don't have yet a story for Service Workers in PyScript that works out of the box and AFAIK we're not planning to have one any time soon, but that might change /cc @ntoll

**Reply by ntoll** [reply]
Yeah... basically, what Andrea says. If upstream Pyodide doesn't support threading, there's not much we can do.

On the use of service workers... if enough folks ask for it, of course we'll investigate, but right now, you're the first.
